### ISSUE TITLE
Code instructions are added to undo stack

### ISSUE THREAD

**Issue by daviddavid** [getcursor/cursor#1229]
After using the code instruction feature (CMD+K), but then later deciding to undo those things with CMD+Z, the code instruction box pops back up and is undone character by character (as it was typed).

Instead, I would expect the code instructions to not be part of my undo stack.

Exception: I assume the reason the instructions are added to the undo stack in the first place, is because you want to be able to undo text you typed there, too. However, everything that happened in the input field should be removed from the undo stack once you're done using it (read accepted or declined changes).

**Comment by allen6u** [comment]
Surprised to see so few people were bothered, I even considered editing the instruction elsewhere and pasting it in as a workaround.

If someone indeed need to modify the instruction through undo, Cursor may consider adding a button or shortcut key to "jump out" (skip) the whole instruction undo part. That could be a global setting or only show up during the instruction undo process.

**Comment by allen7u** [comment]
Surprised to see so few people were bothered, I even considered editing the instruction elsewhere and pasting it in as a workaround.

If someone indeed need to modify the instruction through undo, Cursor may consider adding a button or shortcut key to "jump out" (skip) the whole instruction undo part. That could be a global setting or only show up during the instruction undo process.

**Comment by truell20** [comment]
Hello! Thanks for writing in, feedback like this helps us make the product better :)

This behavior is expected. Open to changing if the current behavior seems broadly unpopular, but I think this is one of those product choices where neither side will get consensus. Many users like having the code instructions in the undo stack, because they'll sometimes want to go back and reprompt the ai after changing their mind about how they want something done.

If the undo's in the prompt bar weren't character by character, would that fix most of your concern?

**Comment by dfeehrer** [comment]
> If the undo's in the prompt bar weren't character by character, would that fix most of your concern?

Voicing my support for this middle ground approach: just treat the entire instruction as one entry on the undo stack. 

I agree instructions should be undoable, but I can't think of any reason I would need to undo them character by character. I find it really frustrating to get stuck undoing a long instruction.

Also fwiw this is one of the longest standing issues we've had with Cursor (submitted this as feedback over email back in Oct 2023), and I hit this ~daily. I'm also surprised more people aren't bothered

**Comment by daviddavid** [comment]
> If the undo's in the prompt bar weren't character by character, would that fix most of your concern?

Yeah, treating the entire prompt as a single undo stack entry would be a lot more sensible and easier to work with, imo. It seems like a minor nitpick, but in an IDE that's centered around prompting, this drove me away from Cursor and back to traditional ways of working. So thanks for picking up this issue, and I'm looking forward to trying out Cursor again in the near future.
